src/jemalloc.c

```diff
@@ -2776,22 +2794,15 @@ free_default(void *ptr) {
 		tsd_t *tsd = tsd_fetch_min();
 		check_entry_exit_locking(tsd_tsdn(tsd));
 
-		tcache_t *tcache;
-		if (likely(tsd_fast(tsd))) {
-			tsd_assert_fast(tsd);
-			/* Unconditionally get tcache ptr on fast path. */
-			tcache = tsd_tcachep_get(tsd);
-			ifree(tsd, ptr, tcache, false);
-		} else {
-			if (likely(tsd_reentrancy_level_get(tsd) == 0)) {
-				tcache = tcache_get(tsd);
-			} else {
-				tcache = NULL;
-			}
+		bool fast = tsd_fast(tsd);
+		tcache_t *tcache = tcache_get_from_ind(tsd,
+		    TCACHE_IND_AUTOMATIC, !fast, /* is_alloc */ false);
+		if (!fast) {
 			uintptr_t args_raw[3] = {(uintptr_t)ptr};
 			hook_invoke_dalloc(hook_dalloc_free, ptr, args_raw);
-			ifree(tsd, ptr, tcache, true);
 		}
+		ifree(tsd, ptr, tcache, !fast);
```

Comment:
We need to repeat the `ifree` call twice -- once with `fast` being statically true, and the other with `fast` being statically false (i.e. inside the if branches). This is key to folding away all the slow-guarded condition checks on the fast path. Likewise -- dropping the `likely` here is potentially optimization-affecting.

Reply:
Right, makes sense. Also repeating the `tcache_get_from_ind()` call. I think now it should work exactly as before.

Reply:
As you were mentioning about the fast path, I realized that I could also unify the fast path so that it calls this centralized function. (And in that case I'll make sure that the assembly stays equally efficient.)